### ISSUE TITLE
PowerTunnel can circumvent ISP torrent throttling but not goodbyeDPI?

### ISSUE THREAD

**Issue by AlawamiAZ** [ValdikSS/GoodbyeDPI#253]
### Operating system / операционная система

Windows 11 22000.493

### Running as service / Запуск программы как сервис

I run it as a regular program / Запускаю программу обычным образом

### Describe the issue / Опишите проблему

Although [PowerTunnel ](https://github.com/krlvm/PowerTunnel) is based on goodbyeDPI, I was surprised that only PowerTunnel seems to circumvent my ISP's bittorrent throttling. goodbyeDPI can unlock all blcoked website so it works but it can't do anything to bittorrent throttling. I tried all presets (1-6) all of them had no effect on bittorrent throttling.

Are there any paramaters that I can try changing to make it able to circumvent my ISPs throttling just like PowerTunnel does. I'm using qBittorrent.

Thanks for the great work on this project.

**Comment by AlawamiAZ** [comment]
Oh great, thanks for clarifiation.. from my understanding -w apply goodbyeDPI on all ports?

Are there any plans to implement something similar to "full chunking mode" of powertunnel? as it seems to be the crucial part of powertunnel that confuses my ISPs throttling mechanism.. unfortunately i'm not familiar with c language so I wasn't able to play around much with goodbyeDPI code.

**Comment by ValdikSS** [comment]
> from my understanding -w apply goodbyeDPI on all ports?

No, it only starts to search for HTTP requests on all additional ports passed with `--port`.

> Are there any plans to implement something similar to "full chunking mode" of powertunnel?

I don't know what is full chunking mode, but this may be already implemented by shrinking the window size. I'll think how it could be implemented.

**Comment by AlawamiAZ** [comment]
Hope you are having a nice day valdik. do you know why goodbyedpi crashes after few seconds of working when compiled with edited line 288 to `WinDivertOpen("true", 0, 0, 0)` ? should't that make it capture all traffic? unfortuantly i'm not versed in C language at all so i've been stepping on a minefield :(

**Comment by krlvm** [comment]
Since modification of the TCP header, which includes Window Size, is impossible without administrator rights and without access to the low-level network, PowerTunnel, upon receiving the packet data, manually splits them into two packets - the first from 0 to N and the second from N to the end, which was recently implemented in GoodbyeDPI as Native Fragmentation, opposed to setting Window Size. What is called "Full Chunking Mode" splits the packet not into 2, but into (packet length / N), where N is the fragment (or chunk) size. I didn't think it had any practical use, this option was never enabled by default and I thought about removing it altogether.


> goodbyedpi crashes after few seconds of working when compiled with edited line 288 to `WinDivertOpen("true", 0, 0, 0)` in `goobyedpi.c` ? i was trying to make it capture all traffic.. it works for first few seconds then crashes with this edit..

Your filter is wrong and does not comply with [WinDivert filter language](https://reqrypt.org/windivert-doc.html#filter_language).
Also, after a very cursory look at the filter, it seems to me that the program only listens for packets that meet certain conditions (the TCP SYN/ACK flags). Perhaps you can achieve what you want by removing the check for ports in `FILTER_STRING_TEMPLATE` and `FILTER_PASSIVE_STRING_TEMPLATE`.

**Comment by AlawamiAZ** [comment]
Thanks krlvm for your intrest. Currently i'm compling with `WinDivertOpen("true", WINDIVERT_LAYER_NETWORK, 0, flags)` and it doesn't cratch and it now defintly catches everything including bittorrent traffic and ICMP..  it still doesn't de-throttle my bittorrent like powertunnel, but i'm looking to modify the fragmentation function of goodbyedpi and see what I can do.

**Comment by krlvm** [comment]
I think you should look into [lines 1291 and 1296](https://github.com/ValdikSS/GoodbyeDPI/blob/b3c9ff84193aeb4b0eb7d571271884dc289d12be/src/goodbyedpi.c#L1291) and rewrite `send_native_fragment` (or create a new option for that) as a loop which sends packets each sized N (except the last one) like [this](https://github.com/krlvm/PowerTunnel/blob/50d1e39eeca471b88e2c20206720f71a863fc02d/littleproxy/src/main/java/org/littleshoot/proxy/impl/ProxyToServerConnection.java#L1536), but without using arrays and with sending packets instantly (it's something I should have done to save memory), but @ValdikSS knows better.
